zio-json/shared/src/main/scala/zio/json/ast/ast.scala

```diff
@@ -56,21 +56,30 @@ sealed abstract class Json { self =>
 
   final def foldUpSome[A](initial: A)(pf: PartialFunction[(A, Json), A]): A = ???
 
-  final def get[A <: Json](cursor: JsonCursor[A]): Either[String, A] =
+  final def get[A <: Json](cursor: JsonCursor[_, A]): Either[String, A] =
     cursor match {
       case JsonCursor.Identity => Right(self)
 
       case JsonCursor.DownField(parent, field) =>
         self.get(parent).flatMap { case Obj(fields) =>
-          Right(fields.find(_._1 == field).map(_._2).getOrElse(Json.Null))
+          fields.find(_._1 == field).map(_._2) match {
+            case Some(x) => Right(x)
+            case None    => Left(s"No such field: '$field'")
+          }
         }
+
       case JsonCursor.DownElement(parent, index) =>
         self.get(parent).flatMap { case Arr(elements) =>
           elements.lift(index).map(Right(_)).getOrElse(Left(s"The array does not have index ${index}"))
         }
 
-      case JsonCursor.FilterType(parent, jsonType) =>
-        self.get(parent).flatMap(jsonType.get(_))
+      case JsonCursor.FilterType(parent, t @ jsonType) =>
+        // TODO: When using 'asJson' on jsonType we can't prove that `A' returned from self.get with
+        //       parent of JsonCursor[_$1] == A
+        self.get(parent).flatMap(x => jsonType.get(x))
```

Comment:
@jdegoes Using `jsonType.asJson: (A => Json)` here doesn’t work, beause `self.get(parent)` with a parent of `JsonCursor[_]` does not return an `A`, which `jsonType` could turn into `Json`.

Reply:
All right. Well, it was worth a try!